Revision of the README.md (#7)
NootInc/WhateverRed: README.md

# WhateverRed

> An AMD iGPU support plugin for Lilu.

Reviewer: Please maintain the link to Lilu

Reviewer: ```suggestion
An AMD iGPU support plugin for [Lilu](https://github.com/acidanthera/Lilu).
```

> This original work's source code is licensed under the `Thou Shalt Not Profit License version 1.0`. See `LICENSE`

Reviewer: The capitalisation and order here matters. It is a legal bind

Reviewer: ```suggestion
The Source Code of this Original Work is licensed under the `Thou Shalt Not Profit License version 1.0`. See `LICENSE`
```

## ⚠️ IMPORTANT ⚠️

> This kernel extension includes remote debugging logging functionality through our custom "NetDbg" software. You must specify a valid NetDbg server IP or disable the functionality as described below.

Reviewer: includes -> contains
debugging logging is bad grammar. Change it to "debug log"

Reviewer: ```suggestion
This kernel extension currently contains remote debug log functionality through our custom "NetDbg" software. You must specify a valid NetDbg server IP or disable the functionality as described below.
```

> WhateverRed NetDbg is a custom remote debugging logging service that sends raw data over TCP. The first version, created in Python by [@NyanCatTW1](https://github.com/NyanCatTW1), was often unstable and primitive. The current release, developed in Rust by project lead [@ChefKissInc](https://github.com/ChefKissInc), is more stable.

Reviewer: Debugging logging again. Seriously? :trollface:

Reviewer: ```suggestion
WhateverRed NetDbg is a custom remote debug log service, which sends raw data over TCP. The first version, created in Python by [@NyanCatTW1](https://github.com/NyanCatTW1), was primitive and often unstable. The current release, developed in Rust by project lead [@ChefKissInc](https://github.com/ChefKissInc), is more stable, and has a lot more features (Like a GUI).
```

> You can find the source code for the service **[here](https://chat.openai.com/chat)**.

Reviewer: What?

Reviewer: ```suggestion

```

Reviewer: ```suggestion
You can find the source code for the service [here](https://github.com/NootInc/WhateverRed-NETDBG).
```

The WhateverRed kext is licensed under BSD-3.

> This project is under active research and development and is not yet functional but close to being so. Its goal is to enable AMD iGPU support for Renoir/Raven(2) and their derivatives, such as Picasso. However, due to the complexity of GPU drivers, especially those without public source code, extended support for non-existent logic is not possible. As a result, you must use Catalina (minimum) to Big Sur (recommended) since some required iGPU logic has been removed in Monterey.

Reviewer: Please add Renoir

Reviewer: ```suggestion
This project is under active research and development and is not yet functional but close to being so. Its goal is to enable AMD iGPU support for Renoir/Raven(2)/Renoir and their derivatives, such as Picasso. However, due to the complexity of GPU drivers, especially those without public source code, extended support for non-existent logic is not possible. As a result, you must use Catalina (minimum) to Big Sur (recommended) since some required iGPU logic has been removed in Monterey.
```

> Injecting kernel extensions is not possible during the OpenCore injection stage.

Reviewer: ```suggestion
Injecting kernel extensions is not possible during the OpenCore injection stage. This is because the prelink injection fails for kernel extensions of this type, as the libraries used for them are not contained in the Boot kernel cache, which is where OpenCore injects kexts to. Instead, they're contained in the Auxiliary kernel cache.
```

## **Introduction**

> We are a team of three working on getting graphics acceleration for AMD iGPUs (Raven/Raven2/Renoir and their derivatives) on hackintoshes (PCs running macOS). Currently, we are fixing the missing and broken logic by patching the existing kexts (AMDRadeonX5000 for GCN 5 and AMDRadeonX6000 for VCN).

Reviewer: The second part is incorrect. Drop it, or add the missing kexts mentioned in the original Telegram message, and remove the word "currently"

Reviewer: ```suggestion
We are a team of three working on getting graphics acceleration for AMD iGPUs (Raven/Raven2/Renoir and their derivatives) on hackintoshes. We are patching the existing kexts in order to achieve this (mixing AMDRadeonX5000 for GCN 5 and AMDRadeonX6000 for VCN 1/2, and replacing AMDRadeonX6000Framebuffer for DCN instead of AMD10000Controller which is DCE).
```

> 

Reviewer: Add a separate FAQ section

Reviewer: ```suggestion
## FAQ

```

> ## **Progress**

Reviewer: ```suggestion
## Current Progress
```

- Controller, accelerator and framebuffer are all attached and enabled.
- GFX engine and SDMA0 engine start up and are able to process command buffers.
- However, the data being fed to the Virtual Memory Page Table (VMPT) is likely incorrect, causing Page Faults in both GFX and ME as soon as WindowServer tries to clear the video memory.
- We are investigating the cause of the Page Faults and will eventually fix it.

> ## **ETA**

Reviewer: ```suggestion
## ETA?
```

Unknown, the project will be completed when it is done.

> ## **Installation**

Reviewer: ```suggestion
## macOS Installation on laptops
```

It may be possible to install macOS on your laptop, but without graphics acceleration and the touchpad may not work.

> ## **EFI Setup**

Reviewer: Redundant section, drop it

Reviewer: (and its contents)

> ## **EFI Setup**

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion

```

> We do not provide support for EFI setup, please follow the AMD section of the Dortania guide and use MacBookPro15,4 as SMBIOS.

Reviewer: ```suggestion

```

## **X5000 vs X6000**

> - X5000: GFX 9, uses VCE + UVD, same VBIOS logic as our device

Reviewer: ```suggestion
- X5000: GFX 9, uses VCE + UVD, DCE 5
```

> - X6000: GFX 10, uses VCN 2, newer VBIOS structure (displays may not be detected properly)

Reviewer: ```suggestion
- X6000: GFX 10, uses VCN 2, DCN 1/2
```

## **Current Status (X5000)**

> - Everything starts out well and the AMD driver even acknowledges the progress.

Reviewer: ```suggestion
- Everything starts out well and the AMD driver even congratulates us with the following message: `[3:0:0]: Controller is enabled, finish initialization`.
```

> - However, as soon as WindowServer tries to zero the video memory, Page Faults occur and cause the SDMA0 engine to freeze, breaking the GFX engine and the entire acceleration process.

Reviewer: ```suggestion
- However, as soon as WindowServer tries to zero the video memory, Page Faults occur and cause the SDMA0 and GFX engines halt, because they fail to read the IB (buffer that contains commands).
```

> - The AMD driver will attempt to reset the GFX/SDMA0 channels every minute, but the Page Faults still occur even after the reset.

Reviewer: ```suggestion
- The AMD driver will attempt to reset the GFX/SDMA0 channels every minute, but the Page Faults persist.
```

> - This results in a black screen for a few minutes, before the machine eventually restarts due to the driver failing to properly reset the channels.

Reviewer: ```suggestion
- This results in a black screen for a few minutes, before the machine eventually restarts due to a watchdog timeout because WindowServer fails to check in success.
```

Note: The X6000 branch has been abandoned for now.

> ###Collaborators:

Reviewer: ```suggestion
### Collaborators:
```
